cpp/src/io/functions.cpp

```diff
@@ -373,54 +358,55 @@ std::unique_ptr<std::vector<uint8_t>> merge_rowgroup_metadata(
+    column_chunks_file_path(op.get_column_chunks_file_path())
 {
-  CUDF_FUNC_RANGE();
-  parquet_writer_options options = parquet_writer_options::builder()
-                                     .compression(op.get_compression())
-                                     .stats_level(op.get_stats_level())
-                                     .int96_timestamps(op.is_enabled_int96_timestamps());
+  writer = make_writer<detail_parquet::writer>(op.get_sink(), op, mr);
+}
 
-  auto state = std::make_shared<pq_chunked_state>();
-  state->wp  = make_writer<detail_parquet::writer>(op.get_sink(), options, mr);
+// Writes table to output
+std::unique_ptr<std::vector<uint8_t>> parquet_writer::write()
+{
+  return writer->write(
+    table, return_filemetadata, column_chunks_file_path, rmm::cuda_stream_default);
+}
 
-  // have to make a copy of the metadata here since we can't really
-  // guarantee the lifetime of the incoming pointer
-  if (op.get_nullable_metadata() != nullptr) {
-    state->user_metadata_with_nullability = *op.get_nullable_metadata();
-    state->user_metadata                  = &state->user_metadata_with_nullability;
-  }
-  state->int96_timestamps = op.is_enabled_int96_timestamps();
-  state->stream           = 0;
-  state->wp->write_chunked_begin(*state);
-  return state;
+// Moves writer unique pointer to object
+parquet_writer& parquet_writer::operator=(parquet_writer&& rhs)
+{
+  writer = std::move(rhs.writer);
+  return *this;
 }
 
-/**
- * @copydoc cudf::io::write_parquet_chunked
- */
-void write_parquet_chunked(table_view const& table, std::shared_ptr<pq_chunked_state> state)
+// Constructor to create parquet chunked writer
+parquet_chunked_writer::parquet_chunked_writer(chunked_parquet_writer_options const& op,
+                                               rmm::mr::device_memory_resource* mr)
 {
-  CUDF_FUNC_RANGE();
-  state->wp->write_chunk(table, *state);
+  writer = make_writer<detail_parquet::writer>(op.get_sink(), op, mr);
 }
 
-/**
- * @copydoc cudf::io::write_parquet_chunked_end
- */
-std::unique_ptr<std::vector<uint8_t>> write_parquet_chunked_end(
-  std::shared_ptr<pq_chunked_state>& state,
-  bool return_filemetadata,
-  const std::string& column_chunks_file_path)
+// Moves writer unique pointer to object
+parquet_chunked_writer& parquet_chunked_writer::operator=(parquet_chunked_writer&& rhs)
 {
-  CUDF_FUNC_RANGE();
-  auto meta = state->wp->write_chunked_end(*state, return_filemetadata, column_chunks_file_path);
-  state.reset();
-  return meta;
+  writer = std::move(rhs.writer);
+  return *this;
+}
+
+// Writes table to output
+void parquet_chunked_writer::write(table_view const& table)
+{
+  writer->write(table, SingleWriteMode::NO);
+}
+
+// Finishes the chunked/streamed write process
+std::unique_ptr<std::vector<uint8_t>> parquet_chunked_writer::write_end(
+  bool return_filemetadata, const std::string& column_chunks_file_path)
+{
+  return writer->write_end(return_filemetadata, column_chunks_file_path);
```

Comment:
What happens if users try to reuse a writer and call `write` after `write_end`?

Reply:
It is writing junk, and rest of the values are null apart from first write. I will add a status var to keep track of it.

Reply:
Added changes to raise error if writer tries to write or flush the data which has been already flushed.